File: src/leetcode/p101to150/LeetCode106ConstructBinaryTreeFromInorderAndPostorderTraversal.java
package leetcode.p101to150;

import java.util.HashMap;
import java.util.Map;

import algorithm.base.TreeNode;

@SuppressWarnings("unused")
public class LeetCode106ConstructBinaryTreeFromInorderAndPostorderTraversal {

    private interface Method {
        TreeNode buildTree(int[] inorder, int[] postorder);
    }

    private static class DFS implements Method {

        private Map<Integer, Integer> mInorderPosMap = new HashMap<>();

        @Override
        public TreeNode buildTree(int[] inorder, int[] postorder) {
            mInorderPosMap.clear();
            for (int i = 0; i < inorder.length; i++) {
                mInorderPosMap.put(inorder[i], i);
            }
            return helper(postorder, 0, 0, postorder.length);
        }

        private TreeNode helper(int[] postorder, int iStart, int pStart, int len) {
            if (len == 0) {
                return null;
            }
            TreeNode root = new TreeNode(postorder[pStart + len - 1]);
            if (len == 1) {
                return root;
            }
            int inorderPos = mInorderPosMap.get(postorder[pStart + len - 1]);
            int lLen = inorderPos - iStart;
            int rLen = len - lLen - 1;

            root.left = helper(postorder, iStart, pStart, lLen);
            root.right = helper(postorder, iStart + lLen + 1, pStart + lLen, rLen);
            return root;
        }
    }
}

File: src/leetcode/p101to150/LeetCode106ConstructBinaryTreeFromInorderAndPostorderTraversal.kt
package leetcode.p101to150

import java.util.HashMap
import basicclass.TreeNode

class LeetCode106ConstructBinaryTreeFromInorderAndPostorderTraversal {

    private var mInorderPosMap: MutableMap<Int, Int> = mutableMapOf()

    fun buildTree(inorder: IntArray, postorder: IntArray): TreeNode? {
        mInorderPosMap = HashMap()
        for (i in inorder.indices) {
            mInorderPosMap[inorder[i]] = i
        }
        return helper(postorder, 0, postorder.size - 1, 0, inorder.size - 1)
    }

    private fun helper(postorder: IntArray,
                       pStart: Int, pEnd: Int,
                       iStart: Int, iEnd: Int): TreeNode? {
        if (iStart > iEnd) {
            return null
        }
        val root = TreeNode(postorder[pEnd])
        if (iStart == iEnd) {
            return root
        }
        val iPos = mInorderPosMap[postorder[pEnd]]!!
        val rightTreeLen = iEnd - iPos

        root.left = helper(postorder, pStart, pEnd - 1 - rightTreeLen, iStart, iPos - 1)
        root.right = helper(postorder, pEnd - rightTreeLen, pEnd - 1, iPos + 1, iEnd)
        return root
    }
}
